staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -745,29 +745,34 @@ func shouldDelegateListOnNotReadyCache(opts storage.ListOptions) bool {
 	return noLabelSelector && noFieldSelector && hasLimit
 }
 
-func (c *Cacher) listItems(ctx context.Context, listRV uint64, key string, pred storage.SelectionPredicate, recursive bool) (listResp, string, error) {
-	if !recursive {
+func (c *Cacher) listItems(ctx context.Context, listRV uint64, key string, opts storage.ListOptions) (listResp, string, error) {
+	if !opts.Recursive {
 		obj, exists, readResourceVersion, err := c.watchCache.WaitUntilFreshAndGet(ctx, listRV, key)
 		if err != nil {
 			return listResp{}, "", err
 		}
 		if exists {
-			return listResp{Items: []interface{}{obj}, ResourceVersion: readResourceVersion}, "", nil
+			return listResp{
+				Items:           []interface{}{obj},
+				ItemCount:       1,
+				ResourceVersion: readResourceVersion,
+			}, "", nil
 		}
 		return listResp{ResourceVersion: readResourceVersion}, "", nil
 	}
-	return c.watchCache.WaitUntilFreshAndList(ctx, listRV, key, pred.MatcherIndex(ctx))
+	return c.watchCache.WaitUntilFreshAndList(ctx, listRV, key, opts)
 }
 
 type listResp struct {
-	Items           []interface{}
+	Items   []interface{}
+	HasMore bool
+	// Return ItemCount but only if pred.Empty()
+	ItemCount       int64
```

Comment:
Looking into this PR, it seems that ItemsCount is always set to len(Items). So what's the point of having it?

Reply:
This is because cache doesn't respect limit, however I expect we want to respect the limit when implementing pagination and exact RV match from cache.

Reply:
We should discuss that more widely - this has a potential to break customers.
Changing that would require input from people like Jordan and David at least.

Reply:
oal is to not break customers and maintain same behavior as etcd currently. When serving requests with `RV=0` the limit is not respected, but I expect that we want to respect the limit for `RV=X`. For that we need itemCount for watch cache.

Reply:
No problem, just wanted to avoid to many changes stacked in one PR. Implementation of limit is pretty complicated.